File: chatgpt-scraper/browser/nodejs/chatgpt.mjs
// ChatGPT scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// function names and emitted field names match verbatim.
//
// Flow:
// - Mint a cloud browser session (CDP WS endpoint).
// - puppeteer-core connects, opens `https://chatgpt.com/?prompt=<text>`.
// - Dismiss the credentials picker, click "send", capture the SSE response
//   from `/backend-anon/f/conversation` and feed it to `parseChatgptStream`.
// - For follow-up prompts, POST `/backend-anon/conversation` from inside
//   the page via in-page fetch (cookies + Cloudflare cookies stay alive).
//
// Public/anonymous content only.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import { randomUUID } from "node:crypto";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 300;

const TRANSIENT_NET_ERRORS = [
  "ERR_TUNNEL_CONNECTION_FAILED",
  "ERR_CONNECTION_CLOSED",
  "ERR_CONNECTION_RESET",
  "ERR_CONNECTION_REFUSED",
  "ERR_TIMED_OUT",
  "ERR_NETWORK_CHANGED",
  "ERR_EMPTY_RESPONSE",
  "ERR_PROXY_CONNECTION_FAILED",
  "Navigation timeout",
  "net::",
];

function isTransientError(err) {
  const msg = String(err?.message ?? err ?? "");
  return TRANSIENT_NET_ERRORS.some((s) => msg.includes(s));
}

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

// Warm up the cloud browser session against the OpenAI marketing origin
// before opening chatgpt.com — establishes Cloudflare cookies on a less-
// defended endpoint so the chatgpt.com TLS handshake doesn't get ripped
// down with ERR_CONNECTION_CLOSED on the first hop.
async function warmupChatgpt(page) {
  try {
    await page.goto("https://chat.openai.com/", { waitUntil: "domcontentloaded", timeout: 30000 });
    await new Promise((r) => setTimeout(r, 2000));
  } catch (_) {}
}

async function withSessionRetry(fn, { retries = 2, label = "chatgpt" } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry: DEFAULT_PROXY_COUNTRY, sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      return await fn(page, browser);
    } catch (e) {
      lastError = e;
      if (!isTransientError(e) && attempt > 0) throw e;
      if (attempt === retries) break;
      const sleepMs = 5000 * Math.pow(2, attempt);
      await new Promise((r) => setTimeout(r, sleepMs));
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`${label}: giving up after ${retries + 1} attempts: ${lastError?.message ?? lastError}`);
}

// ---------------- SSE parser (mirror of the upstream reference's parse_chatgpt_stream) ----------------

export function parseChatgptStream(rawSse) {
  const messages = new Map();
  let conversationId = null;
  let currentId = null;
  let lastO = null;
  let lastP = null;

  const store = (msg) => {
    const id = msg?.id;
    if (!id) return null;
    const parts = msg?.content?.parts ?? [""];
    messages.set(id, {
      role: msg?.author?.role ?? "",
      content: typeof parts[0] === "string" ? parts[0] : "",
    });
    return id;
  };

  const append = (p, o, v) => {
    if (o === "append" && typeof v === "string" && p && p.includes("content/parts/0") && currentId && messages.has(currentId)) {
      const m = messages.get(currentId);
      m.content += v;
      messages.set(currentId, m);
    }
  };

  for (let line of rawSse.split(/\r?\n/)) {
    line = line.trim();
    if (!line.startsWith("data:")) continue;
    const raw = line.slice(5).trim();
    if (raw === "[DONE]") break;
    let data;
    try { data = JSON.parse(raw); } catch { continue; }
    if (!data || typeof data !== "object") continue;

    if (data.type === "input_message") {
      currentId = store(data.input_message ?? {}) ?? currentId;
      conversationId = conversationId ?? data.conversation_id ?? null;
      continue;
    }

    lastO = data.o ?? lastO;
    lastP = data.p ?? lastP;
    const v = data.v;

    if (v && typeof v === "object" && !Array.isArray(v) && "message" in v) {
      currentId = store(v.message) ?? currentId;
      conversationId = conversationId
        ?? v.conversation_id
        ?? v.message?.metadata?.conversation_id
        ?? null;
    } else if (Array.isArray(v)) {
      for (const patch of v) append(patch?.p, patch?.o, patch?.v);
    } else {
      append(lastP, lastO, v);
    }
  }

  let parentMessageId = null;
  const ids = [...messages.keys()];
  for (let i = ids.length - 1; i >= 0; i--) {
    if (messages.get(ids[i]).role === "assistant") {
      parentMessageId = ids[i];
      break;
    }
  }
  const resultMessages = [];
  for (const m of messages.values()) {
    if (m.role && m.content) resultMessages.push({ role: m.role, content: m.content });
  }

  return {
    conversation_id: conversationId,
    parent_message_id: parentMessageId,
    messages: resultMessages,
  };
}

// ---------------- browser orchestration ----------------

async function dismissOverlays(page) {
  try { await page.click("#credentials-picker-container #close", { timeout: 2000 }); } catch (_) {}
  try { await page.click("div[aria-live='polite'] button", { timeout: 2000 }); } catch (_) {}
}

async function sendAndCapture(page, { timeoutMs = 60000 } = {}) {
  const captured = {};
  const handler = async (response) => {
    try {
      const url = response.url();
      if (!url.includes("backend-anon/f/conversation") && !url.includes("backend-anon/conversation")) return;
      const headers = response.headers() || {};
      const ctype = headers["content-type"] || "";
      if (!ctype.includes("event-stream") && !ctype.includes("stream")) return;
      const body = await response.text();
      const req = response.request();
      captured.sse_body = body;
      captured.request_body = req.postData() ?? "";
      captured.request_headers = req.headers();
    } catch (_) {}
  };
  page.on("response", handler);

  await page.waitForSelector("button[data-testid='send-button']", { timeout: 15000 });
  await dismissOverlays(page);
  await page.click("button[data-testid='send-button']");

  const deadline = Date.now() + timeoutMs;
  while (!captured.sse_body && Date.now() < deadline) {
    await new Promise((r) => setTimeout(r, 500));
  }
  page.off("response", handler);
  return captured;
}

// ---------------- scrape functions ----------------

export async function scrapeConversation(prompt) {
  const url = `https://chatgpt.com/?prompt=${encodeURIComponent(prompt)}`;
  return withSessionRetry(async (page) => {
    await warmupChatgpt(page);
    await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
    const captured = await sendAndCapture(page, { timeoutMs: 60000 });
    const parsed = parseChatgptStream(captured.sse_body ?? "");
    const assistant = parsed.messages.filter((m) => m.role === "assistant").map((m) => m.content);
    return assistant.join("\n\n").trim();
  }, { label: "scrape_conversation" });
}

function buildPostRequest(prompt, conversationId, parentMessageId, originalBody, headers) {
  const newBody = { ...originalBody };
  newBody.conversation_id = conversationId;
  newBody.parent_message_id = parentMessageId;
  newBody.messages = [
    {
      id: randomUUID(),
      author: { role: "user" },
      create_time: Date.now() / 1000,
      content: { content_type: "text", parts: [prompt] },
    },
  ];
  return { headers, body: newBody };
}

export async function scrapeConversations(prompts) {
  if (!prompts?.length) return [];
  const firstUrl = `https://chatgpt.com/?prompt=${encodeURIComponent(prompts[0])}`;

  return withSessionRetry(async (page) => {
    let promptIndex = 0;
    const conversations = [];
    await warmupChatgpt(page);
    await page.goto(firstUrl, { waitUntil: "domcontentloaded", timeout: 45000 });
    const captured = await sendAndCapture(page, { timeoutMs: 60000 });

    const parsed = parseChatgptStream(captured.sse_body ?? "");
    let conversationId = parsed.conversation_id;
    let parentMessageId = parsed.parent_message_id;

    if (conversationId) {
      conversations.push({
        conversation_id: conversationId,
        messages: parsed.messages ?? [],
      });
    }

    let originalBody = {};
    try { originalBody = JSON.parse(captured.request_body || "{}"); } catch {}
    const headers = captured.request_headers || {};

    while (promptIndex < prompts.length - 1) {
      promptIndex += 1;
      if (!conversationId) break;
      const req = buildPostRequest(
        prompts[promptIndex], conversationId, parentMessageId ?? "", originalBody, headers,
      );
      const text = await page.evaluate(async ({ url, body, headers }) => {
        const res = await fetch(url, {
          method: "POST",
          headers,
          body: JSON.stringify(body),
          credentials: "include",
        });
        return await res.text();
      }, { url: "https://chatgpt.com/backend-anon/conversation", body: req.body, headers: req.headers });
      const postParsed = parseChatgptStream(text);
      if (postParsed.parent_message_id) parentMessageId = postParsed.parent_message_id;
      if (conversations.length && postParsed.messages?.length) {
        conversations[conversations.length - 1].messages.push(...postParsed.messages);
      }
    }
    return conversations;
  }, { label: "scrape_conversations" });
}

File: chatgpt-scraper/browser/python/chatgpt.py
"""ChatGPT scraper using the official Scrapeless Python SDK + Playwright over CDP.
function names and emitted field names match verbatim, so downstream code
can Flow:
- `client.browser.create()` mints a cloud browser session (CDP WS endpoint).
- Playwright connects over CDP, opens `https://chatgpt.com/?prompt=<text>`.
- We dismiss the credential picker / notice, click the send button, and
  capture the streaming SSE response from `/backend-anon/f/conversation`.
- `parse_chatgpt_stream` (identical to the upstream reference's) folds the SSE event log
  into a `ChatgptConversation` dict.

Public/anonymous content only — ChatGPT requires a real session for any
authenticated history, which this scraper does NOT attempt.
"""

from __future__ import annotations

import asyncio
import json
import os
import time
from typing import Dict, List, Optional, TypedDict
from urllib.parse import quote_plus
from uuid import uuid4

from loguru import logger
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 300

TRANSIENT_NET_ERRORS = (
    "ERR_TUNNEL_CONNECTION_FAILED",
    "ERR_CONNECTION_CLOSED",
    "ERR_CONNECTION_RESET",
    "ERR_CONNECTION_REFUSED",
    "ERR_TIMED_OUT",
    "ERR_NETWORK_CHANGED",
    "ERR_EMPTY_RESPONSE",
    "ERR_PROXY_CONNECTION_FAILED",
    "Navigation timeout",
    "net::",
)

def _is_transient_error(err: Exception) -> bool:
    msg = str(err)
    return any(s in msg for s in TRANSIENT_NET_ERRORS)

async def _warmup_chatgpt(page) -> None:
    """Warm up the cloud browser against chat.openai.com so Cloudflare sets
    its cookies before we hit chatgpt.com. Without this the first TLS
    handshake to chatgpt.com often gets ripped down with ERR_CONNECTION_CLOSED.
    """
    try:
        await page.goto("https://chat.openai.com/", wait_until="domcontentloaded", timeout=30000)
        await asyncio.sleep(2.0)
    except Exception:
        pass

async def _with_session_retry(fn, *, retries: int = 2, label: str = "chatgpt"):
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=DEFAULT_PROXY_COUNTRY, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page()
                return await fn(page)
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("{} attempt {} failed: {}", label, attempt + 1, e)
                if not _is_transient_error(e) and attempt > 0:
                    raise
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
        if attempt < retries:
            await asyncio.sleep(5.0 * (2 ** attempt))
    raise RuntimeError(f"{label}: giving up after {retries + 1} attempts: {last_error}")

# ---------------------------------------------------------------------------
# Types — mirror the upstream reference's TypedDicts verbatim
# ---------------------------------------------------------------------------

class ChatgptMessage(TypedDict):
    role: str
    content: str

class ChatgptConversation(TypedDict):
    conversation_id: str
    messages: List[ChatgptMessage]

# ---------------------------------------------------------------------------
# Scrapeless plumbing
# ---------------------------------------------------------------------------

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

# ---------------------------------------------------------------------------
# SSE parser — verbatim from the upstream reference
# ---------------------------------------------------------------------------

def parse_chatgpt_stream(raw_sse: str) -> Dict:
    """Parse a ChatGPT SSE stream body into structured messages JSON object.

    ChatGPT SSE events come in three shapes (besides "input_message"):
      1. ``v={"message": {...}}`` — seed/finalization of a message object.
      2. ``v=[{p, o, v}, ...]`` — list of JSON-Patch-like operations.
      3. ``v="text"`` with sticky ``p``/``o`` inherited from the previous event.
    """
    messages: Dict[str, dict] = {}
    conversation_id: Optional[str] = None
    current_id: Optional[str] = None
    last_o: Optional[str] = None
    last_p: Optional[str] = None

    def store(msg: dict) -> Optional[str]:
        msg_id = msg.get("id")
        if not msg_id:
            return None
        parts = msg.get("content", {}).get("parts") or [""]
        messages[msg_id] = {
            "role": msg.get("author", {}).get("role", ""),
            "content": parts[0] if isinstance(parts[0], str) else "",
        }
        return msg_id

    def append(path: Optional[str], op: Optional[str], val) -> None:
        if (
            op == "append"
            and isinstance(val, str)
            and path
            and "content/parts/0" in path
            and current_id in messages
        ):
            messages[current_id]["content"] += val

    for line in raw_sse.splitlines():
        line = line.strip()
        if not line.startswith("data:"):
            continue
        raw = line[len("data:"):].strip()
        if raw == "[DONE]":
            break
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        if not isinstance(data, dict):
            continue

        if data.get("type") == "input_message":
            current_id = store(data.get("input_message", {})) or current_id
            conversation_id = conversation_id or data.get("conversation_id")
            continue

        last_o = data.get("o", last_o)
        last_p = data.get("p", last_p)
        v = data.get("v")

        if isinstance(v, dict) and "message" in v:
            current_id = store(v["message"]) or current_id
            conversation_id = (
                conversation_id
                or v.get("conversation_id")
                or v["message"].get("metadata", {}).get("conversation_id")
            )
        elif isinstance(v, list):
            for patch in v:
                append(patch.get("p"), patch.get("o"), patch.get("v"))
        else:
            append(last_p, last_o, v)

    parent_message_id = next(
        (mid for mid, m in reversed(messages.items()) if m["role"] == "assistant"),
        None,
    )
    result_messages: List[ChatgptMessage] = [
        {"role": m["role"], "content": m["content"]}
        for m in messages.values()
        if m["role"] and m["content"]
    ]

    return {
        "conversation_id": conversation_id,
        "parent_message_id": parent_message_id,
        "messages": result_messages,
    }

# ---------------------------------------------------------------------------
# Browser orchestration
# ---------------------------------------------------------------------------

async def _dismiss_overlays(page) -> None:
    """Best-effort: dismiss the credentials picker and notice overlays."""
    try:
        await page.click("#credentials-picker-container #close", timeout=2000)
    except Exception:
        pass
    try:
        await page.click("div[aria-live='polite'] button", timeout=2000)
    except Exception:
        pass

async def _send_and_capture(page, *, timeout_ms: int = 60000) -> Dict[str, str]:
    """Click the send button and capture the SSE conversation response.

    Returns dict with keys: ``sse_body``, ``request_body``, ``request_headers``.
    """
    captured: Dict[str, str] = {}

    async def on_response(response):
        try:
            url = response.url
            if "backend-anon/f/conversation" not in url and "backend-anon/conversation" not in url:
                return
            ctype = (response.headers or {}).get("content-type", "")
            if "event-stream" not in ctype and "stream" not in ctype:
                return
            body = await response.text()
            req = response.request
            captured["sse_body"] = body
            captured["request_body"] = req.post_data or ""
            try:
                captured["request_headers"] = json.dumps(await req.all_headers())
            except Exception:
                captured["request_headers"] = json.dumps(req.headers)
        except Exception as e:
            logger.warning("response capture failed: {}", e)

    page.on("response", lambda r: asyncio.ensure_future(on_response(r)))

    await page.wait_for_selector("button[data-testid='send-button']", timeout=15000)
    await _dismiss_overlays(page)
    await page.click("button[data-testid='send-button']")

    # Wait for SSE body or timeout.
    deadline = asyncio.get_event_loop().time() + (timeout_ms / 1000.0)
    while "sse_body" not in captured and asyncio.get_event_loop().time() < deadline:
        await asyncio.sleep(0.5)
    return captured

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference's exports verbatim
# ---------------------------------------------------------------------------

async def scrape_conversation(prompt: str) -> str:
    """Scrape a single ChatGPT response and return the rendered markdown."""
    url = f"https://chatgpt.com/?prompt={quote_plus(prompt)}"
    logger.info("scraping conversation for prompt: {}", prompt)

    async def _run(page):
        await _warmup_chatgpt(page)
        await page.goto(url, wait_until="domcontentloaded", timeout=45000)
        captured = await _send_and_capture(page, timeout_ms=60000)
        sse_body = captured.get("sse_body", "")
        parsed = parse_chatgpt_stream(sse_body)
        # Return the most recent assistant message as plain text (mirrors
        # the upstream reference's `format="markdown"` content-return contract).
        assistant_msgs = [m["content"] for m in parsed.get("messages", []) if m["role"] == "assistant"]
        content = "\n\n".join(assistant_msgs).strip()
        logger.success("finished scraping ChatGPT for the prompt: {}", prompt)
        return content

    return await _with_session_retry(_run, label="scrape_conversation")

def _build_post_request(
    prompt: str,
    conversation_id: str,
    parent_message_id: str,
    original_body: dict,
    headers: dict,
) -> dict:
    """Build the JSON body and headers for a follow-up POST to /backend-anon/conversation."""
    new_body = dict(original_body)
    new_body["conversation_id"] = conversation_id
    new_body["parent_message_id"] = parent_message_id
    new_body["messages"] = [
        {
            "id": str(uuid4()),
            "author": {"role": "user"},
            "create_time": time.time(),
            "content": {"content_type": "text", "parts": [prompt]},
        }
    ]
    return {"headers": headers, "body": new_body}

async def scrape_conversations(prompt: List[str]) -> List[ChatgptConversation]:
    """Drive a multi-turn anonymous conversation and return the message log."""
    if not prompt:
        return []

    first_url = f"https://chatgpt.com/?prompt={quote_plus(prompt[0])}"

    async def _run(page):
        prompt_index = 0
        conversations: List[ChatgptConversation] = []
        await _warmup_chatgpt(page)
        await page.goto(first_url, wait_until="domcontentloaded", timeout=45000)
        captured = await _send_and_capture(page, timeout_ms=60000)

        sse_body = captured.get("sse_body", "")
        parsed = parse_chatgpt_stream(sse_body)
        conversation_id = parsed.get("conversation_id")
        parent_message_id = parsed.get("parent_message_id")

        if conversation_id:
            conversations.append(
                {
                    "conversation_id": conversation_id,
                    "messages": parsed.get("messages", []),
                }
            )

        try:
            original_body = json.loads(captured.get("request_body") or "{}")
        except json.JSONDecodeError:
            original_body = {}
        try:
            headers = json.loads(captured.get("request_headers") or "{}")
        except json.JSONDecodeError:
            headers = {}

        while prompt_index < len(prompt) - 1:
            prompt_index += 1
            if not conversation_id:
                logger.warning("no conversation_id captured; stopping multi-turn loop")
                break
            req = _build_post_request(
                prompt[prompt_index],
                conversation_id,
                parent_message_id or "",
                original_body,
                headers,
            )
            # POST from within the page so cookies + cloudflare cookies stick.
            text = await page.evaluate(
                """async ({ url, body, headers }) => {
                    const res = await fetch(url, {
                        method: "POST",
                        headers,
                        body: JSON.stringify(body),
                        credentials: "include",
                    });
                    return await res.text();
                }""",
                {
                    "url": "https://chatgpt.com/backend-anon/conversation",
                    "body": req["body"],
                    "headers": req["headers"],
                },
            )
            post_parsed = parse_chatgpt_stream(text)
            if post_parsed.get("parent_message_id"):
                parent_message_id = post_parsed["parent_message_id"]
            if conversations and post_parsed.get("messages"):
                conversations[-1]["messages"].extend(post_parsed["messages"])
        return conversations

    return await _with_session_retry(_run, label="scrape_conversations")

def to_dict(obj):
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj
